Mathlib/Data/Nat/Choose/Cast.lean

```diff
@@ -0,0 +1,52 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+
+! This file was ported from Lean 3 source module data.nat.choose.cast
+! leanprover-community/mathlib commit bb168510ef455e9280a152e7f31673cabd3d7496
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Nat.Choose.Basic
+import Mathlib.Data.Nat.Factorial.Cast
+
+/-!
+# Cast of binomial coefficients
+
+This file allows calculating the binomial coefficient `a.choose b` as an element of a division ring
+of characteristic `0`.
+-/
+
+
+open Nat
+
+variable (K : Type _) [DivisionRing K] [CharZero K]
+
+namespace Nat
+
+theorem cast_choose {a b : ℕ} (h : a ≤ b) : (b.choose a : K) = b ! / (a ! * (b - a)!) := by
+  have : ∀ {n : ℕ}, (n ! : K) ≠ 0 := by simp [cast_ne_zero, factorial_ne_zero]
+  rw [eq_div_iff_mul_eq (mul_ne_zero this this)]
+  rw_mod_cast [← mul_assoc, choose_mul_factorial_mul_factorial h]
+#align nat.cast_choose Nat.cast_choose
+
+theorem cast_add_choose {a b : ℕ} : ((a + b).choose a : K) = (a + b)! / (a ! * b !) := by
+  rw [cast_choose]
+  · rw [add_tsub_cancel_left]
+  · simp
+#align nat.cast_add_choose Nat.cast_add_choose
+
+-- Porting note: changed type ascription to coercion arrow around a - (b - 1)
```

Comment:
I don't think we need a porting note for things where mathport just produced the wrong output.

Reply:
I was worried that there would be a substantial change in the type signature in using the coercion arrow, because the original uses an ascription to $\mathbb N$ and this is to $K$. Nevertheless the note has been removed.

Reply:
The original ascription to ℕ was to make lean insert the coercion to K outside the parens. In lean 4 it seems it is no longer willing to insert it; but the one you inserted manually is identical to the one we wanted it to insert automatically